x-pack/plugin/transform/src/test/java/org/elasticsearch/xpack/transform/transforms/TransformSchedulerTests.java

```diff
@@ -0,0 +1,341 @@
+        TransformScheduler.Listener listener = events::add;
+
+        TransformScheduler transformScheduler = new TransformScheduler(clock, threadPool, SETTINGS);
+        transformScheduler.registerTransform(transformTaskParams, listener);
+        assertThat(
+            transformScheduler.getScheduledTransformTasks(),
+            contains(new ScheduledTransformTask(transformId, frequency, 0L, 0, 5000, listener))
+        );
+        assertThat(events, hasSize(1));
+
+        clock.advanceTimeBy(Duration.ofMillis(5000));
+
+        List<Callable<Void>> concurrentProcessingTasks = new ArrayList<>(10);
+        for (int i = 0; i < 10; ++i) {
+            concurrentProcessingTasks.add(() -> {
+                transformScheduler.processScheduledTasks();
+                return null;
+            });
+        }
+        List<Future<Void>> futures = threadPool.generic().invokeAll(concurrentProcessingTasks);
+        for (Future<Void> future : futures) {
+            future.get();
+        }
+
+        assertThat(events, hasSize(2));
+        assertThat(
+            events,
+            is(
+                equalTo(
+                    List.of(new TransformScheduler.Event(transformId, 0, 0), new TransformScheduler.Event(transformId, 5 * 1000, 5 * 1000))
+                )
+            )
+        );
+    }
+
+    public void testConcurrentModifications() {
+        String transformId = "test-with-fake-clock-concurrent";
+        int frequencySeconds = 5;
+        TimeValue frequency = TimeValue.timeValueSeconds(frequencySeconds);
+        TransformTaskParams transformTaskParams = new TransformTaskParams(transformId, Version.CURRENT, frequency, false);
+        FakeClock clock = new FakeClock(Instant.ofEpochMilli(0));
+        CopyOnWriteArrayList<TransformScheduler.Event> events = new CopyOnWriteArrayList<>();
+
+        TransformScheduler transformScheduler = new TransformScheduler(clock, threadPool, SETTINGS);
+        TransformScheduler.Listener taskModifyingListener = new TransformScheduler.Listener() {
+            private boolean firstTime = true;
+
+            @Override
+            public void triggered(TransformScheduler.Event event) {
+                events.add(event);
+                assertThat(event.getTransformId(), is(equalTo(transformId)));
+                if (firstTime) {
+                    firstTime = false;
+                } else {
+                    transformScheduler.handleTransformFailureCountChanged(transformId, 666);
+                }
+            }
+        };
+        transformScheduler.registerTransform(transformTaskParams, taskModifyingListener);
+        assertThat(
+            transformScheduler.getScheduledTransformTasks(),
+            contains(new ScheduledTransformTask(transformId, frequency, 0L, 0, 5000, taskModifyingListener))
+        );
+        assertThat(events, hasSize(1));
+
+        clock.advanceTimeBy(Duration.ofMillis(5000));
+        transformScheduler.processScheduledTasks();
+    }
```

Comment:
It feels like there is something missing from the test here. There are no checks on the effect of `taskModifyingListener` calling  `transformScheduler.handleTransformFailureCountChanged(transformId, 666);`